llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -1174,6 +1174,86 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
   return IsSigned ? BinaryOperator::CreateNSWAdd(NewMinMax, Add->getOperand(1))
                   : BinaryOperator::CreateNUWAdd(NewMinMax, Add->getOperand(1));
 }
+
+// Try canonicalize max(max(X,C1) binop C2, C3) -> max(X binop C2, C3)
+static Instruction *moveShiftAfterMinMax(IntrinsicInst *II,
```

Comment:
The function name should be updated.
In fact, this fold can be decomposed into two steps:
```
max(max(X,C1) binop C2, C3) -> // Associative laws
max(max(X binop C2, C1 binop C2), C3) -> // Commutative laws
max(X binop C2, max(C1 binop C2, C3)) -> // Constant fold
max(X binop C2, C4)
```

Reply:
`max(X, C1) binop C2 -> max(X binop C2, C1 binop C2)` is not always safe for all binops. You can reuse the helper `leftDistributesOverRight`.